gcp/gcp.go

```diff
@@ -0,0 +1,124 @@
+package gcp
+
+import (
+	"container/list"
+	"github.com/gruntwork-io/cloud-nuke/logging"
+	"golang.org/x/net/context"
+	"golang.org/x/oauth2"
+	google "golang.org/x/oauth2/google"
+	compute "google.golang.org/api/compute/v1"
+	"net/http"
+	"time"
+)
+
+type GcpContext struct {
+	Client  *http.Client
+	Service *compute.Service
+	Project string
+	Regions []*compute.Region
+}
+
+func DefaultContext() (*GcpContext, error) {
+	creds, err := google.FindDefaultCredentials(oauth2.NoContext, compute.ComputeScope)
+	if err != nil {
+		return nil, err
+	}
+	client := oauth2.NewClient(context.Background(), creds.TokenSource)
+
+	service, err := compute.New(client)
+	if err != nil {
+		return nil, err
+	}
+
+	regions, err := service.Regions.List(creds.ProjectID).Do()
+	if err != nil {
+		return nil, err
+	}
+
+	context := &GcpContext{
+		Client:  client,
+		Service: service,
+		Project: creds.ProjectID,
+		Regions: regions.Items,
+	}
+
+	return context, nil
+}
+
+func (ctx *GcpContext) ContainsRegion(region string) bool {
+	for _, r := range ctx.Regions {
+		if r.Name == region {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (ctx *GcpContext) GetAllResources(excludedRegions []string, excludeAfter time.Time) ([]GcpResource, error) {
+	resources := []GcpResource{}
+
+	instances, err := GetAllGceInstances(ctx, excludedRegions, excludeAfter)
+	if err != nil {
+		return nil, err
+	}
+
+	resources = append(resources, instances...)
+
+	return resources, nil
+}
+
+type NukeWorkerResult struct {
+	Resource GcpResource
+	Err      error
+}
+
+func nukeWorker(ctx *GcpContext, resource GcpResource, output chan<- NukeWorkerResult) {
+	err := resource.Nuke(ctx)
+	result := NukeWorkerResult{
+		Resource: resource,
+		Err:      err,
+	}
+	output <- result
+}
+
+func (ctx *GcpContext) NukeAllResources(resources []GcpResource) error {
+	batchSize := 5
+	results := make(chan NukeWorkerResult, 100)
+	resourceList := list.New()
+
+	for _, resource := range resources {
+		resourceList.PushFront(resource)
+	}
+
+	for resourceList.Len() != 0 {
+		i := 0
+		for i < batchSize {
+			element := resourceList.Front()
+			if element == nil {
+				break
+			}
+
+			resourceList.Remove(element)
+
+			go nukeWorker(ctx, element.Value.(GcpResource), results)
+			i += 1
+		}
+
+		for i != 0 {
+			result := <-results
+			if result.Err != nil {
+				logging.Logger.Warnf("could not delete resource: %s: %s Region=%s %s=%s",
+					result.Resource.ResourceName(), result.Resource.Name(), result.Resource.Region(),
+					result.Resource.LocationName(), result.Resource.Location())
+				// TODO certain errors may require a retry
+				//resourceList.PushFront(result.Resource)
```

Comment:
I think rate limiting is done in the library automatically, but still need to look into that a bit more. I have a strong feeling we'll need retries anyway for some reason though, hence why this is a bit hairy.